### DISCUSSION TITLE
`hb-show`: how to turn `kern` on/off with `uniscribe` or `directwrite` shapers

### DISCUSSION THREAD

**Discussion by bobh0303** [harfbuzz/harfbuzz#4855] [Q&A] [answered]
`--features +kern`  and `--features -kern` don't seem to have any impact on `directwrite` and `uniscribe` shapers. From my experiments with `hb-show` it seems that:
- `kern` is always enabled with `uniscribe`
- `kern` is always disabled with `directwrite`

Is this expected?

I'm on Windows 10.0.19045 and using HarfBuzz 9.0.0 release

**Comment by behdad** [comment]
Humm. That's unexpected. Does it work eg. for `liga`? The code is there...

**Reply by bobh0303** [reply]
Similar results for `liga` 

In case it might help, I have tried changing `kern` in the middle of the text, e.g.:
```
hb-view --font-file GentiumPlus-Regular.ttf --text "WA WA" --features +kern[0:2],-kern[3:] -O svg --shaper ot -o wa-ot.svg
```
and it 
- works great for `ot` shaper, 
- kerns entire string in `uniscribe`, 
- generates a mailformed or empty SVG with `directwrite` (sometimes with a message that "Shaping failed").

**Reply by behdad** [reply]
That's all very disturbing. We need someone on Windows to debug this...

**Reply by khaledhosny** [reply, marked as answer]
The `Shaping failed` message with `directwrite` is a bug, should be fixed with https://github.com/harfbuzz/harfbuzz/pull/5114.

The rest is possibly a limitations of how uniscribe works. I believe that for certain scripts it does not allow turning feature on/off.

**Reply by khaledhosny** [reply]
There is https://github.com/miloush/dwbox, one can use it directly to see if DirectWrite supports something but our `directwrite` shaper behaves differently. I don’t have a Windows version it can run on, so I haven’t tested it myself.
